### PR TITLE
feat: [Improvement] Adds FORUM_ENV filter to manage forum environment vars

### PR DESCRIPTION
**Pull request by xitij2000**
The forum service is entirirely configured using environemnt variables, and there are a number of environment variables that are not configurable via tutor currently. This change adds patches in those places allowing additional configuration variables to be passed.

### CODE REVIEW

**Comment by regisb** [comment]
I understand the need to configure the forum via environment variables. I really wish there was a better way to override environment variables than to implement multiple patches, but I thought about it long and hard and did not find a very elegant solution. This PR shows that there is already quite a bit of duplication in the forum templates.

**Comment by regisb** [comment]
I think I got a better idea:

```
FORUM_ENV = tutor.core.hooks.Filter()
FORUM_ENV.add_items([
    ("SEARCH_SERVER", "{{ ELASTICSEARCH_SCHEME }}://{{ ELASTICSEARCH_HOST }}:{{ ELASTICSEARCH_PORT }}"),
    ...
])

@tutor.hooks.Filters.ENV_PATCHES.add()
def _forum_patches(patches):
    k8s_env_patch = ""
    local_env_patch = ""
    for key, value in FORUM_ENV.iterate():
        k8s_env_patch += f"- name: {key}\n  value: {value}\n"
        local_env_patch += f"{key}: {value}\n"
    patches += [("forum-k8s-env", k8s_env_patch), ("forum-local-env", local_env_patch)]
    return patches
```

And then in the k8s templates:

```
env:
    {{ patch("forum-k8s-env")|indent(12) }}
```

(and similar in the docker-compose templates)

With this approach, users would only have to add items to the `FORUM_ENV` filter to define new environment values. Moreover, this would remove a lot of duplication between forum templates.

Please tell me whether this makes sense, I understand it might be a little confusing.

**Comment by xitij2000** [comment]
> I think I got a better idea:
> 
> ```
> FORUM_ENV = tutor.core.hooks.Filter()
> FORUM_ENV.add_items([
>     ("SEARCH_SERVER", "{{ ELASTICSEARCH_SCHEME }}://{{ ELASTICSEARCH_HOST }}:{{ ELASTICSEARCH_PORT }}"),
>     ...
> ])
> 
> @tutor.hooks.Filters.ENV_PATCHES.add()
> def _forum_patches(patches):
>     k8s_env_patch = ""
>     local_env_patch = ""
>     for key, value in FORUM_ENV.iterate():
>         k8s_env_patch += f"- name: {key}\n  value: {value}\n"
>         local_env_patch += f"{key}: {value}\n"
>     patches += [("forum-k8s-env", k8s_env_patch), ("forum-local-env", local_env_patch)]
>     return patches
> ```
> 
> And then in the k8s templates:
> 
> ```
> env:
>     {{ patch("forum-k8s-env")|indent(12) }}
> ```
> 
> (and similar in the docker-compose templates)
> 
> With this approach, users would only have to add items to the `FORUM_ENV` filter to define new environment values. Moreover, this would remove a lot of duplication between forum templates.
> 
> Please tell me whether this makes sense, I understand it might be a little confusing.

I think this makes a lot more sense! I was just reluctant to add new config to the main project. 
Here, instead of using a patch though, wouldn't it be better to just directly iterate over `FORUM_ENV` in the templates?

**Comment by xitij2000** [comment]
Just realised what I proposed won't work since the values won't be rendered. So I've implemented your approach.

**Comment by xitij2000** [comment]
I generated the config with and without this change and here is the diff:

tutor-forum-pr-24

So nothing that should break anything.

File: tutorforum/plugin.py [hunk 1]
```diff
@@ -21,6 +21,18 @@
...
+    "SEARCH_SERVER": "{{ ELASTICSEARCH_SCHEME }}://{{ ELASTICSEARCH_HOST }}:{{ ELASTICSEARCH_PORT }}",
+    "MONGODB_AUTH":
+     "{% if MONGODB_USERNAME and MONGODB_PASSWORD %}{{ MONGODB_USERNAME}}:{{ MONGODB_PASSWORD }}@{% endif %}",
+    "MONGODB_HOST": "{{ MONGODB_HOST|forum_mongodb_host }}",
+    "MONGODB_PORT": "{{ MONGODB_PORT }}",
+    "MONGODB_DATABASE": "{{ FORUM_MONGODB_DATABASE }}",
+    "MONGOID_AUTH_SOURCE": "{{ MONGODB_AUTH_SOURCE }}",
+    "MONGOID_AUTH_MECH": "{{ MONGODB_AUTH_MECHANISM|auth_mech_as_ruby }}",
+    "MONGOID_USE_SSL": "{{ 'true' if MONGODB_USE_SSL else 'false' }}",
+}
```
Comment (xitij2000) [inline tutorforum/plugin.py:34]: @regisb I decided to make it a dict instead of a list of tuples since this makes overriding individual environment variables simpler.

**Comment by gabor-boros** [comment]
👍 @xitij2000

File: tutorforum/hooks.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+"""
+These hooks are stored in a separate module. If they were included in plugin.py, then
+the tutor-forum hooks would be created in the context of some other plugin that imports
+them.
+"""
+from tutor.core.hooks import Filter
```
Comment (regisb) [inline tutorforum/hooks.py:6]: I think you need to add `from __future__ import annotations` to get this to run on Python 3.8.

File: tutorforum/plugin.py [hunk 3]
```diff
@@ -112,6 +124,23 @@ def forum_mongodb_host(host: str) -> str:
...
 
 
+@FORUM_ENV.add(priority=tutor_hooks.priorities.HIGH)
+def _add_base_forum_env(forum_env: dict) -> dict[str, str]:
+    forum_env.update(FORUM_ENV_BASE)
+    return forum_env
+
+
+@tutor_hooks.Filters.ENV_PATCHES.add()
+def _forum_env_patches(patches):
```
Comment (regisb) [inline tutorforum/plugin.py:134]: This is a valid approach but it's uncommon, would you mind adding a comment to explain what is going on?

**Review by regisb — CHANGES_REQUESTED** [review]
I only have a couple minor comments, otherwise I think this is top notch. @ghassanmas what do you think?

File: tutorforum/plugin.py (hunk 3 above)
Comment (xitij2000) [reply tutorforum/plugin.py:137]: Sure, I've added comments and docstrings to explain what's going on here.

File: tutorforum/plugin.py [hunk 4]
```diff
@@ -21,6 +21,18 @@
     },
 }
 
+FORUM_ENV_BASE: dict[str,str] = {
+    "SEARCH_SERVER": "{{ ELASTICSEARCH_SCHEME }}://{{ ELASTICSEARCH_HOST }}:{{ ELASTICSEARCH_PORT }}",
```
Comment (ghassanmas) [inline tutorforum/plugin.py:25]: This line got me an error when trying to test it locally with python 3.8  (which is included in `setup.py` )
When I tried to quickly test it I got the following error 
```
    FORUM_ENV_BASE: dict[str,str] = {
TypeError: 'type' object is not subscriptable
```
I think you might just need to import the type

**Comment by ghassanmas** [comment]
This looks interesting change, I have quickly tested and provided a comment above... I will keep an eye on this and expect me to test it locally with docker/compose.  But I wouldn't test it with k8s as it's still a black box to me. 

A curious quesiotn I have; What would happen if opreator would use the k8s/local patch directly? like do we expect them to do so? _I suppose it might be the case if I have different configruation between local and k8s_

Another thing that also make the situation more obscure to me is that it might be the case that we run _k8s locally_ hence
-  openedx/wg-devops/issues/4 

But this I guess then something related to tutor core naming as always naming might not reflect the meaning :-)

File: tutorforum/plugin.py (hunk 4 above)
Comment (xitij2000) [reply tutorforum/plugin.py:27]: Iv'e added a `from __future__ import annotations` to the plugin.py file as well, which should fix this.

**Comment by xitij2000** [comment]
> This looks interesting change, I have quickly tested and provided a comment above... I will keep an eye on this and expect me to test it locally with docker/compose. But I wouldn't test it with k8s as it's still a black box to me.
> 
> A curious quesiotn I have; What would happen if opreator would use the k8s/local patch directly? like do we expect them to do so? _I suppose it might be the case if I have different configruation between local and k8s_
> 
> Another thing that also make the situation more obscure to me is that it might be the case that we run _k8s locally_ hence
> 
>     * [How to run Open edX on a local Kubernetes cluster? openedx/wg-devops#4](https://github.com/openedx/wg-devops/issues/4)
> 
> 
> But this I guess then something related to tutor core naming as always naming might not reflect the meaning :-)

I can't think for a good reason right now for someone to want different environment variables for k8s vs local. However, they can just use the patches in that case as you mentioned. 

I imagine people will generally only deploy an instance using local or k8s and not both from a single config, so if they wanted they can just copy over the config, change the required variables/plugins/settings and deploy to local and k8s from different configs.

I think running k8s on your local computer is a kind of an edge case! I've tested tutor again minikube and I currently run a k8s cluster on a secondary mini-pc exactly for these kinds of tests. I've had success so far, but it definitely takes some effort setting up.

**Comment by ghassanmas** [comment]
I have tested this locally and it worked as expected, the services init, craeted a thread..etc. The only thing I can think of, is just ot update the readme, and second regarding the order/priority of the patches, my thinking is that we have know three pathces that can override each other: 
1. The forum built-in patch _that is filled_ from tutor config 
2. The forum env base 
3. The local/k8s variants 

My understanding and I think that the logical is tha should be loaded in that order, but I haven't tested that yet, looking through the code I see that 2 and 3 has priority high, so I think we should be safe... _unless we get into a weired edge case if opreator are using both 2 and 3 and 3 get loaded before 2_

**Comment by xitij2000** [comment]
> I have tested this locally and it worked as expected, the services init, craeted a thread..etc. The only thing I can think of, is just ot update the readme, and second regarding the order/priority of the patches, my thinking is that we have know three pathces that can override each other:
> 
>     1. The forum built-in patch _that is filled_ from tutor config
> 
>     2. The forum env base
> 
>     3. The local/k8s variants
> 
> 
> My understanding and I think that the logical is tha should be loaded in that order, but I haven't tested that yet, looking through the code I see that 2 and 3 has priority high, so I think we should be safe... _unless we get into a weired edge case if opreator are using both 2 and 3 and 3 get loaded before 2_

I've updated the readme file with some info about how this hook can be used. 


As for the priority. I think that's definitely a potential issue. I'm not sure how to fully avoid it, and perhaps it can't be avoided in all cases. Do you think it's worth warning about this in the readme? I've run into such edge cases in the past and it was hard to diagnose, but perhaps also hard to list all such edge cases.

File: README.rst [hunk 5]
```diff
@@ -34,6 +34,29 @@ Configuration
...
+
+To add, or modify environment variables that are supplied to the forum service,
+you can use the ``FORUM_ENV`` hook.
+
+To add or modify a environment variable, update the corresponding entry in the
+``FORUM_ENV`` dict as follows::
+
+	from tutorforum.hooks import FORUM_ENV
+
+	@FORUM_ENV.add()
```
Comment (regisb) [inline README.rst:48]: minor: indentation looks weird from here

**Review by regisb — APPROVED** [review]
LGTM! Just a minor formatting issue

File: README.rst (hunk 5 above)
Comment (xitij2000) [reply README.rst]: Fixed, and added code highlighting as well.

**Comment by xitij2000** [comment]
> LGTM! Just a minor formatting issue

@regisb Just checking if this is OK to merge now.

**Review by regisb — APPROVED** [review]
LGTM! As the maintainer of this repo, I'll let @ghassanmas take care of squashing and merging.
